backend/src/crd/controller/scheduledworkflow/util/cron_schedule.go

```diff
@@ -42,79 +42,83 @@ func NewCronSchedule(cronSchedule *swfapi.CronSchedule) *CronSchedule {
 	// time.
-	effectiveLastJobEpoch := defaultStartEpoch
-	if lastJobEpoch != nil {
+	effectiveLastJobTime := defaultStartTime
+	if lastJobTime != nil {
 		// Last job epoch takes first precedence.
-		effectiveLastJobEpoch = *lastJobEpoch
+		effectiveLastJobTime = lastJobTime.Time
 	} else if s.StartTime != nil {
 		// Start time takes second precedence.
-		effectiveLastJobEpoch = s.StartTime.Unix()
+		effectiveLastJobTime = s.StartTime.Time
 	}
-	return effectiveLastJobEpoch
+	return effectiveLastJobTime
 }
 
-func (s *CronSchedule) getNextScheduledEpoch(lastJobEpoch int64) int64 {
-	return s.getNextScheduledEpochImp(lastJobEpoch,
-		true, 0 /* nowEpoch doesn't matter when catchup=true */)
+func (s *CronSchedule) getNextScheduledEpoch(lastJobTime time.Time, location *time.Location) int64 {
+	return s.getNextScheduledEpochImp(lastJobTime,
+		true /* nowEpoch doesn't matter when catchup=true */, time.Unix(0, 0), location)
 }
 
-func (s *CronSchedule) getNextScheduledEpochImp(lastJobEpoch int64, catchup bool, nowEpoch int64) int64 {
+func (s *CronSchedule) getNextScheduledEpochImp(lastJobTime time.Time, catchup bool, nowTime time.Time, location *time.Location) int64 {
 	schedule, err := cron.Parse(s.Cron)
 	if err != nil {
 		// This should never happen, validation should have caught this at resource creation.
 		log.Errorf("%+v", wraperror.Errorf(
 			"Found invalid schedule (%v): %v", s.Cron, err))
-		return math.MaxInt64
+		return time.Unix(1<<63-62135596801, 0).Unix()
 	}
 
-	startEpoch := lastJobEpoch
-	if s.StartTime != nil && s.StartTime.Unix() > startEpoch {
-		startEpoch = s.StartTime.Unix()
+	startTime := lastJobTime
+	if s.StartTime != nil && s.StartTime.Time.After(startTime) {
+		startTime = s.StartTime.Time
 	}
-	result := schedule.Next(time.Unix(startEpoch, 0).UTC()).Unix()
 
-	var endTime int64 = math.MaxInt64
+	result := schedule.Next(startTime.In(location))
+	var endTime time.Time = time.Unix(1<<63-62135596801, 0)
+	// math.int64 max will break the comparison.
+	// Examle playground https://play.golang.org/p/LERg0aq2mU6
+	// Max date https://stackoverflow.com/questions/25065055/what-is-the-maximum-time-time-in-go
 	if s.EndTime != nil {
-		endTime = s.EndTime.Unix()
+		endTime = s.EndTime.Time
 	}
-	if endTime < result {
-		return math.MaxInt64
+
+	if endTime.Before(result) {
+		return time.Unix(1<<63-62135596801, 0).Unix()
 	}
 
 	// When we need to catch up with schedule, just run schedules one by one.
 	if catchup == true {
-		return result
+		return result.Unix()
 	}
 
 	// When we don't need to catch up, find the last schedule we need to run
 	// now and skip others in between.
 	next := result
-	var nextNext int64
+	var nextNext time.Time
 	for {
-		nextNext = schedule.Next(time.Unix(next, 0).UTC()).Unix()
-		if nextNext <= nowEpoch && nextNext <= endTime {
+		nextNext = schedule.Next(next)
+		if (nextNext.Before(nowTime) || nextNext.Equal(nowTime)) && (nextNext.Before(endTime) || nextNext.Equal(endTime)) {
```

Comment:
Will !xxx.After(guy) be easier to understand?

Reply:
Yes, updated to use it instead.